[17.0][MIG] account_invoice_triple_discount: Migration to 17.0

ramiadavid opened the pull request.

chienandalu committed:
account_invoice_triple_discount: New module

pedrobaeza committed:
account_invoice_triple_discount: Some improvements

xavierjimenez committed:
account_invoice_triple_discount

rschnapka committed:
Translated using Weblate (German)

Currently translated at 100,0% (4 of 4 strings)

Translation: account-invoicing-11.0/account-invoicing-11.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-11-0/account-invoicing-11-0-account_invoice_triple_discount/de/

Nikul-Chaudhary committed:
[MIG] account_invoice_triple_discount v11 to v12

oca-travis committed:
[UPD] Update account_invoice_triple_discount.pot

OCA-git-bot committed:
[ADD] icon.png

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (4 of 4 strings)

Translation: account-invoicing-12.0/account-invoicing-12.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-12-0/account-invoicing-12-0-account_invoice_triple_discount/pt_BR/

benwillig committed:
[MIG] account_invoice_triple_discount to v13

ferran-S73 committed:
[IMP] account_invoice_triple_discount: black, isort, prettier

OCA-git-bot committed:
[14.0][MIG] account_invoice_triple_discount: Migration to V14

dreispt committed:
[FIX] account_invoice_triple_discount: Developer Mode was required to use discounts

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (4 of 4 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_invoice_triple_discount/it/

OCA-git-bot committed:
[IMP] update dotfiles

ramiadavid committed:
[MIG] account_invoice_triple_discount: Migration to 15.0

Commit:
[UPD] Update account_invoice_triple_discount.pot

OCA-git-bot committed:
[UPD] README.rst

jabelchi committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (4 of 4 strings)

Translation: account-invoicing-15.0/account-invoicing-15.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-15-0/account-invoicing-15-0-account_invoice_triple_discount/ca/

ramiadavid committed:
[FIX] account_invoice_triple_discount: fix precision when compute tax with discount

(cherry picked from commit 5afdb213c8f87be17bb21401a45cbf05a3548aa9)

ramiadavid committed:
[MIG] account_invoice_triple_discount: Migration to 16.0

Commit:
[UPD] Update account_invoice_triple_discount.pot

OCA-git-bot committed:
[UPD] README.rst

ramiadavid committed:
[FIX][16.0] account_invoice_triple_discount : Error when print reports

Commit:
[UPD] Update account_invoice_triple_discount.pot

OCA-git-bot committed:
account_invoice_triple_discount 16.0.1.0.1

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/hr/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/pt_BR/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/es/

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/

SirAionTech committed:
[FIX] account_invoice_triple_discount: Discount rounding

OCA-git-bot committed:
[BOT] post-merge updates

ramiadavid committed:
[IMP] account_invoice_triple_discount: pre-commit auto fixes

ramiadavid committed:
[MIG] account_invoice_triple_discount: Migration to 17.0

sbidoul commented:
/ocabot migration account_invoice_triple_discount

grindtildeath committed:
[REF] account_invoice_triple_discount: Consolidate discount in std field

Until now, the triple discount feature did use the standard Odoo field
as the first discount field, adding only extra fields for the second
and the third discount. This implied we had to override any function
using discount to consider the other discounts properly.

By adding an extra field discount1 to store the first discount, it allows
to redefine the standard discount field to a computed field that will
consolidate the triple discount from the other fields, and avoid the
need to redefine anything relying on the discount field as it will
already consider the triple discounts.

nb-obs commented:
Hello guys, As I see module is not available on website. Is there any update on this module? Any kind of help is appreciated.

lmignon committed:
[FIX] account_invoice_triple_discount: Fix tax recompute on closed fiscal year

When computing the taxes and the totals on account.move.line it's required to prevent calls to the write method on account.move or triggering recompute of unexpected fields while we assign a temporary value to the discount field to use the aggregated one bedore calling the base implementation. Otherwise the update of the discount field will invalidate totals and ... on related accoun.move and trigger calls to the write method of account.move. Any call to the write method will trigger a set of check to see if the balance is still right or check if you try to write on a record prior to the fiscal year close date or .... At the end, every time you try to access to the taxes on the records, some errors could be unexpeced raised due to the trick a temporarily changing the discount value of the field while computing others fields.

In the future we should uses dedicated fields for 3 discount fields and make the native discount field a computed one with readonly=False. This approach would simplify a lot the code. See https://github.com/OCA/account-invoicing/pull/1638

lmignon committed:
[IMP] account_invoice_triple_discount: Improves tests

Uses CommonCase to take benefit of proper env initialization and avoid useless tacking records creation

salvorapi commented:
@sbidoul please, can you merge this on 17.0 branch ? thank you, Salvo.

Boxnia reviewed (approved):
LGTM

sbidoul commented:
> @sbidoul please, can you merge this on 17.0 branch ? thank you, Salvo.

I'm not a contributor nor an expert on this module, so I'll let those who know take care of that.

salvorapi reviewed (approved):
LGTM

legalsylvain committed:
[IMP] account_invoice_triple_discount: Simplify account.move.line algorithm

[FIX] account_invoice_triple_discount: populate discount1 field to avoid inconsistency if database contains account.move.line with not null discount before the installation of the module

[REF] account_invoice_triple_discount: Add GRAP as coauthor of the module

[FIX] account_invoice_triple_discount: Manage Correct creation of account move lines, where 'discount' field is set

legalsylvain reviewed (changes requested):
Hi. I think that the huge simplication introduced by https://github.com/OCA/account-invoicing/pull/1638 should be integrated in the version 17.

@ramiadavid could you review it ? 

thanks

ramiadavid commented:
@legalsylvain I'm waiting for the PR to be done

legalsylvain commented:
> @legalsylvain I'm waiting for the PR to be done

your review is welcome !

ramiadavid committed:
Add test

rousseldenis commented:
@ramiadavid @legalsylvain @Boxnia IMHO, the proposal to refactor triple discount modules in v16 should be reviewed and attended before merging v17 : https://github.com/OCA/account-invoicing/pull/1638 and https://github.com/OCA/sale-workflow/pull/2885

grindtildeath committed:
Write discount into discount 1 if not defined

In case this module is installed and sale module is also installed
but not sale_triple_discount, or in case sale_triple_discount is not
updated to its latest version, make sure that setting discount, without
having discount1 defined, will not mess with the triple discounts.

ferran-S73 commented:
@ramiadavid please add commit https://github.com/OCA/account-invoicing/commit/2a2b975deed6b7c4af9dd966e3caa3cc3adf9629 which was merged in v16

antoniocanovas reviewed (approved):
Functional review OK !!

legalsylvain commented:
> @ramiadavid @legalsylvain @Boxnia IMHO, the proposal to refactor triple discount modules in v16 should be reviewed and attended before merging v17 : https://github.com/OCA/account-invoicing/pull/1638 and https://github.com/OCA/sale-workflow/pull/2885

indeed. merge of #1638 is still pending.

legalsylvain committed:
[FIX] account_invoice_fixed_discount, account_invoice_triple_discount:
- set modules as incompatible in the manifest. (For the time being, there is only a informative message in the README of the module account_invoice_fixed_discount)
- Run separated tests for both modules.

legalsylvain committed:
[FIX] account_invoice_triple_discount: remove digits on overloaded field 'discount' as the field is now a technical field that should not have limitation. For exemple in the default case where digits precision of Discount is 2, if discounts are 05%, 09% and 13% the main discount is 24.7885 % (and not 24.79)

legalsylvain committed:
[REF] account_invoice_triple_discount: Improve speed computation of the field 'discount' and reduce complexity

legalsylvain committed:
[FIX] account_invoice_triple_discount: do not hide column that are hidden, because display_discount is False

legalsylvain committed:
[FIX] account_invoice_triple_discount: discounts fields was hidden, because was t-if display_discount (that is False).
[IMP] account_invoice_triple_discount: display only discount column with not null discount

BurkhalterY commented:
Hi,
There is 4 approvals. Do you think we could merge?

legalsylvain commented:
@ramiadavid could you cherry pick :   
- https://github.com/OCA/account-invoicing/pull/1840
- https://github.com/OCA/account-invoicing/pull/1938

Ojce done, we can merge.

legalsylvain commented on account_invoice_triple_discount/migrations/16.0.2.0.0/pre-migrate.py:
```diff
@@ -0,0 +1,47 @@
+# Copyright 2024 Camptocamp SA
```
Could you remove this obsolète file ?

legalsylvain reviewed (approved):
A remark inline. Otherwise, LGTM. Thanks !

eugenios73 commented:
Hi @ramiadavid ! Can you solve the tests errors, please?

ivilata commented:
@ramiadavid, it would be great if you could address the pending errors. They happened a while ago so the logs are no longer available and there's no way to check what the errors were, sadly.

If you don't feel like taking up the task, I may try and clone your branch, add the changes that @legalsylvain suggested and check what the outcome is. Thanks anyway for you contributions!

ramiadavid commented:
Hi, @ivilata this module causes problems with tests in other modules, but I don't really think they affect its functionality outside of the test:

`account_invoice_pricelist`: In this case, I can fix the error by modifying the discount field, limiting it to 13 decimal places. Because when calculating the total discount, python gives results like 9.999999999999998 instead of 10.0

`account_global_discount`: In this case, I'm not sure how to fix it. Instead, I would need to modify the test that fails in this module:
The error occurs when trying to modify a field that is now read-only.

        with Form(self.invoice) as invoice_form:
            with invoice_form.invoice_line_ids.edit(0) as line_form:
                line_form.discount = 20

Can i add this:

    @api.model
    def fields_get(self, allfields=None, attributes=None):
        res = super().fields_get(allfields, attributes)
        if self.env.context.get("test_account_global_discount") and res.get("discount"):
            res["discount"]["readonly"] = False
        return res

But I don't know if it's right to do this kind of things.

ramiadavid committed:
[FIX] account_invoice_triple_discount: fix errors with tests from other modules

legalsylvain commented:
> account_invoice_pricelist: In this case, I can fix the error by modifying the discount field, limiting it to 13 decimal places. Because when calculating the total discount, python gives results like 9.999999999999998 instead of 10.0

Please use assertAlmostEqual

https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertAlmostEqual


    with Form(self.invoice) as invoice_form:
        with invoice_form.invoice_line_ids.edit(0) as line_form:
            line_form.discount = 20

can be replaced by 

    with Form(self.invoice) as invoice_form:
        with invoice_form.invoice_line_ids.edit(0) as line_form:
            # Makes test working if account_invoice_triple_discount is installed
            if "discount_1" in line_form._fields:
                line_form.discount_1 = 20
            else:
                line_form.discount = 20


(pseudo code, maybe some adaptation is required)

ivilata commented:
Thanks @ramiadavid for the explanation, maybe @legalsylvain's suggestion (thank you too!) helps avoid the issue with other modules, it looks cleaner to me than the `fields_get()` hack.

ivilata commented:
@legalsylvain, @ramiadavid, @rousseldenis: PR #2114 supersedes this one and it does include the refactoring suggested above. It passes all tests and merging it may unlock PR OCA/purchase-workflow#2230 as well. If you had some time to review it, that would be great. Thanks a lot! :slightly_smiling_face:
